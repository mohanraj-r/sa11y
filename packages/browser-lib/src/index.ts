/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { base, recommended, full } from '@sa11y/preset-rules';
export { exceptionListFilter } from '@sa11y/format';
export { assertAccessible } from '@sa11y/assert';

export const nameSpace = 'sa11y';
