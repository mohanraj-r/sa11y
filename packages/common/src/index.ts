/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { A11yConfig, AxeResults, axeRuntimeExceptionMsgPrefix, axeVersion, getViolations } from './axe';
export { AssertFunction, Options, WDIOBrowser } from './wdio';
export { errMsgHeader, ExceptionList } from './format';
