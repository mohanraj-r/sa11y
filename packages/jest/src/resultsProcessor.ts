/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AggregatedResult, AssertionResult, SerializableError } from '@jest/test-result/build/types';
import { TestResult } from '@jest/reporters';
import { A11yError } from '@sa11y/format';
import { buildFailureTestResult } from '@jest/test-result';

type FailureDetail = {
    error?: A11yError;
};

// Map of test suite name to test results
const consolidatedErrors = new Map<string, AssertionResult[]>();

// suite.name = [Sa11y WCAG-`2.0`-Level-`A` SC-`1.1.1` `image-alt`]: `testSuite.testFilePath`
// test.name = CSS Selectors[0]
// test.failure.message =
//  first line: Accessibility issue found:`description`
//  rest: Help URL, CSS Selectors, List of tests `test.name`, Sa11y Help URL

function addTestResult(testResult: AssertionResult, testSuite: TestResult) {
    // const sa11yResults = [] as TestResult[];
    // Copy test suite meta-data and clear test results
    // const sa11yTestSuite = { ...testSuite, testResults: [] };
    const suiteName = testSuite.testFilePath.substring(testSuite.testFilePath.lastIndexOf('/') + 1);
    testResult.failureDetails.forEach((failure) => {
        let error = (failure as FailureDetail).error;
        // If using circus test runner https://github.com/facebook/jest/issues/11405#issuecomment-843549606
        if (error === undefined) error = failure as A11yError;
        if (error.name === A11yError.name) {
            // TODO : What happens if ever there are multiple failureDetails? Are there ever?
            error.violations.forEach((violation) => {
                violation.nodes.forEach((a11yError) => {
                    const suiteKey = `[Sa11y ${violation.id}]: ${suiteName}`;
                    if (!consolidatedErrors.has(suiteKey)) consolidatedErrors.set(suiteKey, []);
                    consolidatedErrors.get(suiteKey)?.push({
                        ...testResult,
                        fullName: `${a11yError.target[0]}`, // CSS Selector
                        failureMessages: [
                            `Accessibility issue found: ${violation.help}
 Help: ${violation.helpUrl.split('?')[0]}
 CSS Selectors: "${a11yError.target.join('; ')}"
 Tests: "${testResult.fullName}"
 More info: https://sfdc.co/a11y-jest`,
                        ],
                    });
                });
            });
            // Don't report the failure twice
            testResult.status = 'disabled';
            // Suites with only a11y errors should be marked as passed
            testSuite.numFailingTests -= 1;
            // TODO (fix): Remove sa11y msg from test suite message
            //  ANSI codes and test names in suite message makes it difficult
            // testResult.failureMessages = [];
            // testResult.failureDetails = [];
            testSuite.failureMessage = '';
        }
    });
}

/**
 * Custom results processor for a11y results. Only affects JSON results file output.
 * To be used with jest cli options --json --outputFile
 *  e.g. jest --json --outputFile jestResults.json --testResultsProcessor node_modules/@sa11y/jest/dist/resultsProcessor.js
 * Ref: https://jestjs.io/docs/configuration#testresultsprocessor-string
 *  - Mapping of AggregatedResult to JSON format to https://github.com/facebook/jest/blob/master/packages/jest-test-result/src/formatTestResults.ts
 */
export function resultsProcessor(results: AggregatedResult): AggregatedResult {
    // const sa11yResults = [] as TestResult[];

    results.testResults
        .filter((testSuite) => testSuite.numFailingTests > 0)
        .forEach((testSuite) => {
            // Copy test suite meta-data and clear test results
            // const sa11yTestSuite = { ...testSuite, testResults: [] };

            testSuite.testResults
                .filter((testResult) => testResult.status === 'failed')
                .forEach((testResult) => addTestResult(testResult, testSuite));

            // if (sa11yTestSuite.testResults.length > 0) {
            //     sa11yResults.push(sa11yTestSuite);
            // }
        });

    // results.testResults.push(...sa11yResults);

    consolidatedErrors.forEach((testResults, suiteKey) => {
        const sa11ySuite = buildFailureTestResult(suiteKey, new Error() as SerializableError);
        sa11ySuite.testResults = testResults;
        results.testResults.push(sa11ySuite);
    });

    return results;
}

// The processor must be a node module that exports a function
module.exports = resultsProcessor;
