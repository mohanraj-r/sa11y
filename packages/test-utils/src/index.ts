/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export {
    audioURL,
    domWithA11yIssues,
    domWithA11yIssuesBodyID,
    domWithNoA11yIssues,
    htmlFileWithA11yIssues,
    htmlFileWithNoA11yIssues,
    a11yIssuesCount,
    shadowDomID,
    videoURL,
} from './test-data';
export { beforeEachSetup, cartesianProduct, checkA11yError } from './utils';
