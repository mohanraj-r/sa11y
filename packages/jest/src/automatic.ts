/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { AxeResults } from '@sa11y/common';
import { getViolationsJSDOM } from '@sa11y/assert';
import { A11yError } from '@sa11y/format';

/**
 * Options for Automatic checks to be passed to {@link registerSa11yAutomaticChecks}
 */
export type AutoCheckOpts = {
    runAfterEach?: boolean;
    cleanupAfterEach?: boolean;
    // TODO (feat): add support for result consolidation across all tests for a test run
    //  and optional exclusion of selected tests
    // consolidateResults?: boolean;
    // excludeTests?: string[];
};

/**
 * Default options when {@link registerSa11yAutomaticChecks} is invoked
 */
const defaultAutoCheckOpts: AutoCheckOpts = {
    runAfterEach: true,
    cleanupAfterEach: true,
};

/**
 * Run accessibility check on each element node in the body using {@link toBeAccessible}
 * @param cleanup - boolean indicating if the DOM should be cleaned up after
 */
export async function automaticCheck(cleanup = false): Promise<void> {
    const violations: AxeResults = [];
    // Create a DOM walker filtering only elements (skipping text, comment nodes etc)
    const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_ELEMENT);
    let currNode = walker.firstChild();
    try {
        while (currNode !== null) {
            // TODO (spike): Use a logger lib with log levels selectable at runtime
            // console.log(
            //     `♿ [DEBUG] Automatically checking a11y of ${currNode.nodeName}
            //      for test "${expect.getState().currentTestName}"
            //      : ${expect.getState().testPath}`
            // );
            violations.push(...(await getViolationsJSDOM(currNode)));
            currNode = walker.nextSibling();
        }
    } finally {
        if (cleanup) document.body.innerHTML = ''; // remove non-element nodes
        A11yError.checkAndThrow(violations);
    }
}

/**
 * Register accessibility checks to be run automatically after each test
 * @param opts - Options for automatic checks {@link AutoCheckOpts}
 */
export function registerSa11yAutomaticChecks(opts: AutoCheckOpts = defaultAutoCheckOpts): void {
    if (opts.runAfterEach) {
        console.log('♿ Registering sa11y checks to be run automatically after each test');
        afterEach(async () => {
            await automaticCheck(opts.cleanupAfterEach);
        });
    }
}
