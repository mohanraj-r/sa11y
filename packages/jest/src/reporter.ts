/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { Reporter, ReporterOnStartOptions, Test } from '@jest/reporters';
import { AggregatedResult, TestCaseResult } from '@jest/test-result/build/types';
import { A11yError } from '@sa11y/format';
import { AxeResults } from '@sa11y/common';
import { writeFileSync } from 'fs';

/**
 * Custom Jest Reporter to output consolidated a11y issues as JSON file.
 * Notes:
 * 1. Reporter and tests are executed in their own sandboxes - objects cannot be mutated
 *  or propagated across. Only way to communicate is via exception thrown from tests
 *  which is accessible as part of the test instance in reporter.
 * 2. Values passed to reporter hooks are readonly. Modifications e.g. to test instance
 *  pass/fail status do not take effect in Jest.
 * 3. JSDOM environment is not available in reporter. Hence invoking automatic checks
 *  e.g. from within reporter hooks is not possible.
 */
export default class Sa11yReporter implements Reporter {
    static testCount = 0;
    static violations: AxeResults = [];

    /**
     * Triggered at end of each test
     * @param test - readonly instance containing test path, context and duration
     * @param testCaseResult - readonly instance containing result of test run
     * TODO(Debug): Not able to get this hook to trigger from jest with
     *  default test runner https://github.com/facebook/jest/issues/11405
     */
    onTestCaseResult(test: Test, testCaseResult: TestCaseResult): void {
        if (testCaseResult.status === 'failed') {
            testCaseResult.failureDetails.forEach((failure) => {
                // TODO: instanceof check doesn't work for A11yError
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore TS2339: Property 'name' does not exist on type 'Object'.
                if (failure instanceof Object && failure.name === A11yError.name) {
                    const violations = (failure as A11yError).violations;
                    Sa11yReporter.violations.push(...violations);
                    Sa11yReporter.testCount += violations.length;
                }
            });
        }
    }

    /**
     * Triggered after all tests have been executed.
     */
    onRunComplete(): void {
        console.log('testCount', Sa11yReporter.testCount);
        console.log('violations', Sa11yReporter.violations);
        writeFileSync('sa11y.json', JSON.stringify(Sa11yReporter.violations, null, 2));
    }

    // Required methods in the Reporter interface - currently not being used in this reporter
    /* eslint-disable @typescript-eslint/no-empty-function */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    onRunStart(results: AggregatedResult, options: ReporterOnStartOptions): void | Promise<void> {}
    getLastError(): void | Error {}
    /* eslint-enable @typescript-eslint/no-empty-function */
}
