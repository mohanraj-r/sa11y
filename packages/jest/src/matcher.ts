/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { matcherHint, printReceived } from 'jest-matcher-utils';
import { adaptA11yConfig } from './setup';
import { A11yCheckableContext, assertAccessible } from '@sa11y/assert';
import { A11yError, Options } from '@sa11y/format';
import { A11yConfig, recommended } from '@sa11y/preset-rules';

// Type def for custom jest a11y matchers
// Ref: https://jestjs.io/docs/en/expect.html#expectextendmatchers
declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace jest {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        interface Matchers<R> {
            toBeAccessible(config?: A11yConfig): Promise<CustomMatcherResult>;
        }
    }
}

const matcherHintMsg = `expected document to have no accessibility violations but found`;
const expectedMsg = `0 issues`;
const formatOptions: Options = {
    a11yViolationIndicator: 'â­•',
    helpUrlIndicator: 'ðŸ”—',
    highlighter: printReceived,
};

/**
 * Jest expect matcher to check DOM for accessibility issues
 * @param received - DOM or HTML Element to be tested for accessibility. Defaults to current DOM.
 * @param config - A11yConfig to be used to test for accessibility. Defaults to recommended.
 */
export async function toBeAccessible(
    received: A11yCheckableContext = document,
    config: A11yConfig = recommended
): Promise<jest.CustomMatcherResult> {
    let isAccessible = true;
    // a11yError can be 'undefined' when used with `not` matcher and there is no error
    let a11yError: A11yError = new A11yError([]);
    let receivedMsg = expectedMsg;

    // TODO (Improvement): Can we detect if this is invoked async and error if not ?
    try {
        await assertAccessible(received, adaptA11yConfig(config));
    } catch (e) {
        if (e instanceof A11yError) {
            a11yError = e;
            isAccessible = false;
            receivedMsg = `${a11yError.length} issues`;
        } else {
            throw e;
        }
    }

    return {
        pass: isAccessible,
        message: (): string =>
            matcherHint(
                `${toBeAccessible.name}: ${matcherHintMsg} ${receivedMsg}: \n\n ${a11yError.format(formatOptions)}`,
                `${a11yError.length} issues`,
                expectedMsg
            ),
    };
}
