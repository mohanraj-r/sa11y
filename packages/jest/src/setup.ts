/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { toBeAccessible } from './matcher';
import { A11yConfig } from '@sa11y/preset-rules';

/**
 *
 */
export type AutoCheckOpts = {
    runAfterEach: boolean;
    cleanupAfterEach: boolean;
    excludeTests: string[];
};

const defaultAutoCheckOpts: AutoCheckOpts = {
    runAfterEach: false,
    cleanupAfterEach: false,
    excludeTests: [],
};

/**
 * Options to be passed on to {@link registerSa11yMatcher}
 */
export type Sa11yOpts = {
    autoCheckOpts: AutoCheckOpts;
};

const defaultSa11yOpts: Sa11yOpts = {
    autoCheckOpts: defaultAutoCheckOpts,
};

/**
 * Register accessibility helpers toBeAccessible as jest matchers
 */
export function registerSa11yMatcher(opts: Sa11yOpts = defaultSa11yOpts): void {
    if (expect !== undefined) {
        expect.extend({ toBeAccessible });
    } else {
        throw new Error(
            "Unable to find Jest's global expect." +
                '\nPlease check you have added @sa11y/jest correctly to your jest configuration.' +
                '\nSee https://github.com/salesforce/sa11y/tree/master/packages/jest#readme for help.'
        );
    }

    if (opts.autoCheckOpts.runAfterEach) {
        afterEach(async () => {
            // TODO (spike): Is there a better way to walk the DOM ?
            //  https://developer.mozilla.org/en-US/docs/Web/API/Node/firstChild
            while (document.body.firstChild) {
                try {
                    await expect(document.body.firstChild).toBeAccessible();
                } finally {
                    if (opts.autoCheckOpts.cleanupAfterEach) {
                        document.body.removeChild(document.body.firstChild);
                    }
                }
            }
        });
    }
}

/**
 * Customize sa11y preset rules specific to JSDOM
 */
export function adaptA11yConfig(config: A11yConfig): A11yConfig {
    // TODO: Is it worth checking if we are running in jsdom before modifying config ?
    //  Ref: https://github.com/jsdom/jsdom/issues/1537#issuecomment-229405327
    // const runningInJSDOM = navigator.userAgent.includes('Node.js') || navigator.userAgent.includes('jsdom');
    // if (!runningInJSDOM) return config;
    return {
        ...config,
        rules: {
            // Disable color-contrast as it is doesn't work for JSDOM and might affect performance
            //  https://github.com/dequelabs/axe-core/issues/595
            //  https://github.com/dequelabs/axe-core/blob/develop/doc/examples/jsdom/test/a11y.js
            'color-contrast': { enabled: false },
        },
    };
}
