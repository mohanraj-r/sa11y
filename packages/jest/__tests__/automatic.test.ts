/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { setup } from '../src';
import * as automatic from '../src/automatic';
import { automaticCheck, registerSa11yAutomaticChecks } from '../src/automatic';
import { beforeEachSetup, checkA11yError, domWithA11yIssues, domWithNoA11yIssues } from '@sa11y/test-utils';

describe('automatic checks registration', () => {
    afterAll(() => {
        jest.restoreAllMocks();
        process.env.SA11Y_AUTO = '';
        process.env.SA11Y_CLEANUP = '';
    });

    const registerAutomaticMock = jest.spyOn(automatic, 'registerSa11yAutomaticChecks');

    it('should run when called directly without setup', () => {
        registerSa11yAutomaticChecks(); // exercising default args for code cov
        expect(registerAutomaticMock).toHaveBeenCalledWith();
    });

    it('should not run by default via setup', () => {
        setup();
        expect(registerAutomaticMock).toHaveBeenCalledWith({
            runAfterEach: false,
            cleanupAfterEach: false,
        });
    });

    it('should run when opted in via setup', () => {
        setup({ autoCheckOpts: { runAfterEach: true } });
        expect(registerAutomaticMock).toHaveBeenCalledWith({
            runAfterEach: true,
            cleanupAfterEach: false,
        });
    });

    it('should run when opted in with env vars', () => {
        process.env.SA11Y_AUTO = '1';
        setup();
        expect(registerAutomaticMock).toHaveBeenCalledWith({
            runAfterEach: true,
            cleanupAfterEach: false,
        });

        process.env.SA11Y_CLEANUP = '1';
        setup();
        expect(registerAutomaticMock).toHaveBeenCalledWith({
            runAfterEach: true,
            cleanupAfterEach: true,
        });
    });
});

describe('automatic checks call', () => {
    beforeEach(beforeEachSetup);
    // Note: "expect"s are in the helper method "checkA11yError"
    /* eslint-disable jest/expect-expect */

    it('should not raise a11y issues for DOM without a11y issues', async () => {
        document.body.innerHTML = domWithNoA11yIssues;
        await automaticCheck(); // throws no error
    });

    it('should raise a11y issues for DOM with a11y issues', async () => {
        document.body.innerHTML = domWithA11yIssues;
        // Note: cleanup required to prevent domWithA11yIssues being checked again after
        // the test as part of the afterEach hook that was setup in the previous
        // describe block
        await automaticCheck(true).catch((e) => checkA11yError(e));
    });
    /* eslint-enable jest/expect-expect */

    it('should not cleanup DOM by default', async () => {
        document.body.innerHTML = domWithNoA11yIssues;
        await automaticCheck();
        expect(document.body.innerHTML).not.toBe('');
        // TODO (refactor): replace with a regex check on DOM content
        // expect(domWithNoA11yIssues.replace(/\s/g, '')).toContain(document.body.innerHTML.replace(/\s/g, ''));
    });

    it('should cleanup DOM when opted in', async () => {
        document.body.innerHTML = domWithNoA11yIssues;
        await automaticCheck(true);
        expect(document.body.innerHTML).toBe('');
    });
});
