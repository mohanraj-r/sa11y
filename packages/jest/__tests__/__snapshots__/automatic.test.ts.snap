// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`automatic checks call should raise a11y issues for DOM with a11y issues 1`] = `
"2 accessibility issues found
 * (link-name) Links must have discernible text: #dom-with-issues > a
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/link-name

* (link-name) Links must have discernible text: #second-link
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/link-name"
`;
